libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1411 @@
+                     flatpak_remote_ref_action_ref (value));
+
+  g_ptr_array_sort (squashed_ref_actions, sort_flatpak_remote_ref_actions);
+
+  return squashed_ref_actions;
+}
+
+/* Given a hash table of filenames to FlatpakRemoteRefActionsFile, hoist
+ * the underlying pointer array of remote ref actions and make that the value
+ * of the new hash table.
+ *
+ * This will make the hash table suitable for passing to
+ * eos_updater_util_squash_remote_ref_actions */
+GHashTable *
+eos_updater_util_hoist_flatpak_remote_ref_actions (GHashTable *ref_actions_file_table)
+{
+  g_autoptr(GHashTable) hoisted_ref_actions_table = g_hash_table_new_full (g_str_hash,
+                                                                           g_str_equal,
+                                                                           g_free,
+                                                                           (GDestroyNotify) g_ptr_array_unref);
+  gpointer key, value;
+  GHashTableIter iter;
+
+  g_hash_table_iter_init (&iter, ref_actions_file_table);
+
+  while (g_hash_table_iter_next (&iter, &key, &value))
+    {
+      FlatpakRemoteRefActionsFile *ref_actions_file = value;
+
+      g_hash_table_insert (hoisted_ref_actions_table,
+                           g_strdup (key),
+                           g_ptr_array_ref (ref_actions_file->remote_ref_actions));
+    }
+
+  return g_steal_pointer (&hoisted_ref_actions_table);
+}
+
+/* Examine each of the remote ref action lists in ref_action_table
+ * and squash them down into a list where only one action is applied for
+ * each flatpak ref (the latest one) */
+GHashTable *
+eos_updater_util_squash_remote_ref_actions (GHashTable *ref_actions_table)
+{
+  gpointer key;
+  gpointer value;
+  GHashTableIter hash_iter;
+  GHashTable *squashed_ref_actions_table = g_hash_table_new_full (g_str_hash,
+                                                                  g_str_equal,
+                                                                  g_free,
+                                                                  (GDestroyNotify) g_ptr_array_unref);
+
+  g_hash_table_iter_init (&hash_iter, ref_actions_table);
+
+  while (g_hash_table_iter_next (&hash_iter, &key, &value))
+    {
+      GPtrArray *unsquashed_remote_ref_actions = value;
+      GPtrArray *squashed_remote_ref_actions = squash_ref_actions_ptr_array (unsquashed_remote_ref_actions);
+
+      g_hash_table_insert (squashed_ref_actions_table,
+                           g_strdup (key),
+                           squashed_remote_ref_actions);
+    }
+
+  return squashed_ref_actions_table;
+}
+
+/* Take the most recent (eg, last) entry in each ref actions list in the
+ * ref_actions_table and return a hash table mapping each ref action filename
+ * to a counter indicating how up to date we are on the system for that
+ * file */
+GHashTable *
+eos_updater_util_remote_ref_actions_to_expected_progresses (GHashTable *ref_actions_table)
```

Comment:
Is this function used? From some quick grepping it doesn’t seem to be. Unless it’s used in the tests, which I haven’t loaded for this review.

Reply:
Hum, seems like it isn't used anymore. I'll remove it.